ecs-init/gpu/nvidia_gpu_manager.go

```diff
@@ -0,0 +1,218 @@
+	GetDriverVersion() (string, error)
+	DetectGPUDevices() error
+	SaveGPUState() error
+}
+
+// NvidiaGPUManager is used as a wrapper for NVML APIs and implements GPUManager
+// interface
+type NvidiaGPUManager struct {
+	DriverVersion       string
+	NvidiaDockerVersion string
+	GPUIds              []string
+}
+
+const (
+	// NvidiaGPUDeviceFilePattern is the pattern of GPU device files on the instance
+	NvidiaGPUDeviceFilePattern = "/dev/nvidia*"
+	// GPUInfoDirPath is the directory where gpus and driver info are saved
+	GPUInfoDirPath = "/var/lib/ecs/gpu"
+	// NvidiaGPUInfoFilePath is the file path where gpus and driver info are saved
+	NvidiaGPUInfoFilePath = "/var/lib/ecs/gpu/nvidia-gpu-info.json"
+	// FilePerm is the file permissions for gpu info json file
+	FilePerm = 0700
+)
+
+// ErrNoGPUDeviceFound is thrown when it is not a ECS GPU instance
+var ErrNoGPUDeviceFound = errors.New("No GPU device files found in the instance")
+
+// NewNvidiaGPUManager is used to obtain NvidiaGPUManager handle
+func NewNvidiaGPUManager() GPUManager {
+	return &NvidiaGPUManager{}
+}
+
+// Setup is used for setting up gpu information in the instance
+func (n *NvidiaGPUManager) Setup() error {
+	err := n.DetectGPUDevices()
+	if err != nil {
+		if err == ErrNoGPUDeviceFound {
+			return nil
+		}
+		return errors.Wrapf(err, "setup failed")
+	}
+	err = n.Initialize()
+	if err != nil {
+		return errors.Wrapf(err, "setup failed")
+	}
+	defer n.Shutdown()
+	version, err := n.GetDriverVersion()
+	if err != nil {
+		return errors.Wrapf(err, "setup failed")
+	}
+	n.DriverVersion = version
+	gpuIds, err := n.GetGPUDeviceIDs()
+	if err != nil {
+		return errors.Wrapf(err, "setup failed")
+	}
+	n.GPUIds = gpuIds
+	err = n.SaveGPUState()
+	if err != nil {
+		return errors.Wrapf(err, "nvidia gpu manager: setup failed")
+	}
+	return nil
+}
+
+// DetectGPUDevices checks if GPU devices are present in the instance
+func (n *NvidiaGPUManager) DetectGPUDevices() error {
+	matches, err := MatchFilePattern(NvidiaGPUDeviceFilePattern)
+	if err != nil {
+		return errors.Wrapf(err, "nvidia gpu manager: detecting GPU devices failed")
+	}
+	if matches == nil {
+		return ErrNoGPUDeviceFound
+	}
+	return nil
+}
+
+var MatchFilePattern = FilePatternMatch
+
+func FilePatternMatch(pattern string) ([]string, error) {
+	return filepath.Glob(pattern)
+}
+
+// Initialize is for initlializing nvidia's nvml library
+func (n *NvidiaGPUManager) Initialize() error {
+	err := InitializeNVML()
+	if err != nil {
+		return errors.Wrapf(err, "nvidia gpu manager: error initializing nvidia nvml")
+	}
+	return nil
+}
+
+var InitializeNVML = InitNVML
+
+func InitNVML() error {
+	return nvml.Init()
+}
+
+// Shutdown is for shutting down nvidia's nvml library
+func (n *NvidiaGPUManager) Shutdown() error {
+	err := ShutdownNVML()
+	if err != nil {
+		return errors.Wrapf(err, "nvidia gpu manager: error shutting down nvidia nvml")
+	}
+	return nil
+}
+
+var ShutdownNVML = ShutdownNVMLib
```

Comment:
Seems we don't have a consistent pattern for the naming, if would be better if we follow the patter something like below, that way we will be able to distinguish them by name:
```
var NVMLShutdown = Shutdown
func Shutdown() ...

var NVMLGetDriverVersion = GetDriverVersion
func GetDriverVersion()...
```